### PR TITLE
PLAT-10530 Presence service implementation

### PR DESCRIPTION
**Pull request by symphony-youness**
### Ticket
[PLAT-10530](https://perzoinc.atlassian.net/browse/PLAT-10530)

### Description
- implementation
- tests
- added typehint to the ServiceFactory.get_health_service function
- As I understood, the symphony developers documentation is enough for thisservice: https://developers.symphony.com/restapi/reference#get-presence

### Checklist
- [x] Referenced a ticket in the PR title and in the corresponding section
- [x] Filled properly the description and dependencies, if any
- [x] Unit tests updated or added
- [x] Docstrings added or updated
- [x] Updated the documentation in [docs folder](../docs)

### CODE REVIEW

**Comment by symphony-youness** [comment]
I have 2 things that I wanted to discuss:
- For the tests, I will merge all test resource json files for a specific service in one, with the test name as the key to the json dict , for example, regarding the presence service tests. Instead of having, resources > presence_response > test1.json, test2.json ...
I will have one json file named, `presence_response.json` containing:
```json
{"test1": {
  "aField": "aThing",
  "anotherField": 1234
},
  "test2": {
    "otherField": "something"
  }
}
```

- In the generated V2Presence model, they say that for the category attribute (the response): 
`Possible values are:   - UNDEFINED   - AVAILABLE   - BUSY   - DO_NOT_DISTURB   - ON_THE_PHONE   - BE_RIGHT_BACK   - N_A_MEETING   - AWAY   - OUT_OF_OFFICE   - OFF_WORK   - OFFLINE` 
In java, the PresenceStatus Enum doesn't contain the `DO_NOT_DISTURB`, `OFFLINE` and `UNDEFINED`. Is that done on purpose, idk ?

**Comment by symphony-elias** [comment]
> I have 2 things that I wanted to discuss:
> 
> * For the tests, I will merge all test resource json files for a specific service in one, with the test name as the key to the json dict , for example, regarding the presence service tests. Instead of having, resources > presence_response > test1.json, test2.json ...
>   I will have one json file named, `presence_response.json` containing:
> 
> ```json
> {"test1": {
>   "aField": "aThing",
>   "anotherField": 1234
> },
>   "test2": {
>     "otherField": "something"
>   }
> }
> ```
> 
> * In the generated V2Presence model, they say that for the category attribute (the response):
>   `Possible values are:   - UNDEFINED   - AVAILABLE   - BUSY   - DO_NOT_DISTURB   - ON_THE_PHONE   - BE_RIGHT_BACK   - N_A_MEETING   - AWAY   - OUT_OF_OFFICE   - OFF_WORK   - OFFLINE`
>   In java, the PresenceStatus Enum doesn't contain the `DO_NOT_DISTURB`, `OFFLINE` and `UNDEFINED`. Is that done on purpose, idk ?

For the tests, I don't  mind
For the presence status values, the enum is used only for the set presence method: https://developers.symphony.com/restapi/reference#set-presence -> it mentions only a subset of all possible values (the one put in the Java BDK enum PresenceStatus)

File: symphony/bdk/core/service/presence/presence_service.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+        :param feed_id: The presence feed id to be deleted.
+        :return: The id of the deleted presence feed.
+        `Delete Presence Feed <https://developers.symphony.com/restapi/reference#delete-presence-feed"`_
+        """
+        string_id = await self._presence_api.v1_presence_feed_feed_id_delete_post(
+            session_token=await self._auth_session.session_token, feed_id=feed_id)
+        return string_id.id
+
+    async def set_user_presence(self, user_id, status, soft):
```
Comment (symphony-youri) [inline symphony/bdk/core/service/presence/presence_service.py]: can't we type the status with the enum here? (and soft too, and user_id)

File: symphony/bdk/core/service/presence/presence_service.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        last read.
+
+        :return: Presence feed Id
+        `Create Presence Feed <https://developers.symphony.com/restapi/reference#create-presence-feed"`_
+        """
+        string_id = await self._presence_api.v1_presence_feed_create_post(
+            session_token=await self._auth_session.session_token)
+        return string_id.id
+
+    async def read_presence_feed(self, feed_id):
```
Comment (symphony-youri) [inline symphony/bdk/core/service/presence/presence_service.py]: can we add type hints to this method and the one after

File: symphony/bdk/core/service/presence/presence_service.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        AWAY)
+        :return: Presence info of the calling user.
+        `Set Presence <https://developers.symphony.com/restapi/reference#set-presence"`_
+        """
+        presence_status: V2PresenceStatus = V2PresenceStatus(category=status.value)
+        return await self._presence_api.v2_user_presence_post(session_token=await self._auth_session.session_token,
+                                                              presence=presence_status,
+                                                              soft=soft)
+
+    async def create_presence_feed(self):
```
Comment (symphony-youri) [inline symphony/bdk/core/service/presence/presence_service.py]: we probably did it the same in Java but is the presence feed an OBO service?

File: symphony/bdk/core/service_factory.py [hunk 4]
```diff
@@ -119,12 +121,25 @@ def get_datafeed_loop(self) -> AbstractDatafeedLoop:
...
         )
 
-    def get_health_service(self):
+    def get_health_service(self) -> HealthService:
         """Returns a fully initialized HealthService
 
         :return: a new HealthService instance
         """
-        return HealthService(AgentSystemApi(self._agent_client), SignalsApi(self._agent_client))
+        return HealthService(
```
Comment (symphony-youri) [inline symphony/bdk/core/service_factory.py:129]: do we have different formatting config or was it not formatted before?

File: symphony/bdk/core/service/presence/presence_service.py (hunk 3 above)
Comment (thibauult) [reply symphony/bdk/core/service/presence/presence_service.py]: I do think yes, from https://developers.symphony.com/restapi/reference#obo-enabled-endpoints

File: symphony/bdk/core/service_factory.py (hunk 4 above)
Comment (symphony-youness) [reply symphony/bdk/core/service_factory.py:129]: It was not formatted before.

File: symphony/bdk/core/service/presence/presence_service.py (hunk 3 above)
Comment (symphony-elias) [reply symphony/bdk/core/service/presence/presence_service.py]: Yes we did the same in Java: https://github.com/finos/symphony-bdk-java/blob/main/symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/presence/OboPresenceService.java

File: symphony/bdk/core/service/presence/presence_service.py (hunk 2 above)
Comment (symphony-youness) [reply symphony/bdk/core/service/presence/presence_service.py]: added arguments and returns type hints to all functions

**Comment by symphony-youri** [comment]
> I have 2 things that I wanted to discuss:
> 
> * For the tests, I will merge all test resource json files for a specific service in one, with the test name as the key to the json dict , for example, regarding the presence service tests. Instead of having, resources > presence_response > test1.json, test2.json ...
>   I will have one json file named, `presence_response.json` containing:
> 
> ```json
> {"test1": {
>   "aField": "aThing",
>   "anotherField": 1234
> },
>   "test2": {
>     "otherField": "something"
>   }
> }
> ```
> 
> * In the generated V2Presence model, they say that for the category attribute (the response):
>   `Possible values are:   - UNDEFINED   - AVAILABLE   - BUSY   - DO_NOT_DISTURB   - ON_THE_PHONE   - BE_RIGHT_BACK   - N_A_MEETING   - AWAY   - OUT_OF_OFFICE   - OFF_WORK   - OFFLINE`
>   In java, the PresenceStatus Enum doesn't contain the `DO_NOT_DISTURB`, `OFFLINE` and `UNDEFINED`. Is that done on purpose, idk ?

Isn't it better to keep one response file per test, to see them as recorded responses for each API call?
Otherwise we would have to write logic to parse this file and say what is used by which test
What if you want to reuse a response file for different tests?

File: tests/resources/presence_response/presence_payload.json [hunk 5]
```diff
@@ -0,0 +1,39 @@
+{
```
Comment (symphony-youri) [inline tests/resources/presence_response/presence_payload.json:1]: it becomes more difficult to me to see the exact payload returned by a ws

File: symphony/bdk/core/service/presence/presence_service.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+from symphony.bdk.gen.pod_model.v2_presence_status import V2PresenceStatus
+from symphony.bdk.gen.pod_model.v2_user_presence import V2UserPresence
+
+
+class PresenceStatus(Enum):
+    """The list of all possible values for the presence status.
+    Set Presence <https://developers.symphony.com/restapi/reference#set-presence>`_
+    """
+
+    def _generate_next_value_(name, start, count, last_values):
```
Comment (symphony-elias) [inline symphony/bdk/core/service/presence/presence_service.py]: This is fine but you could also use the `name` field to get the enum name as String, like `PresenceStatus.AVAILABLE.name` will return `"AVAILABLE"`.

**Review by symphony-youri — APPROVED** [review]

File: symphony/bdk/core/service/presence/presence_service.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+        :param soft: If true, the user's current status is taken into consideration. If the user is currently OFFLINE,
+        the user's presence will still be OFFLINE, but the new presence will take effect when the user comes online. If
+        the user is currently online, the user's activity state will be applied to the presence if applicable. (e.g. if
+        you are setting their presence to AVAILABLE, but the user is currently idle, their status will be represented as
+        AWAY)
+        :return: Presence info of the calling user.
+        `Set Presence <https://developers.symphony.com/restapi/reference#set-presence"`_
+        """
+        presence_status: V2PresenceStatus = V2PresenceStatus(category=status.value)
```
Comment (symphony-elias) [inline symphony/bdk/core/service/presence/presence_service.py]: Regarding my comment on PresenceStatus enum, you could use `status.name` instead.

File: symphony/bdk/core/service/presence/presence_service.py (hunk 6 above)
Comment (symphony-youness) [reply symphony/bdk/core/service/presence/presence_service.py]: sure, I just thought that using value can be more concise. I will change it and remove the overridden method

File: tests/resources/presence_response/presence_payload.json (hunk 5 above)
Comment (symphony-youness) [reply tests/resources/presence_response/presence_payload.json:1]: For me it's a choice. I thought that centring it like this avoided adding json strings in tests making them bulkier. If you wish to do the same as java, tell me and I will.

Reuse file for multiple tests:
I plan to check if adding scope=module for the fixture makes use of the same json for all tests. (we wouldn't open the file for each test)

**Comment by symphony-youness** [comment]
For some reason I wrongly added the !fixup at the begging of the last commit's message. Will clean that during squash/merge on monday.

File: tests/resources/presence_response/presence_payload.json (hunk 5 above)
Comment (symphony-youri) [reply tests/resources/presence_response/presence_payload.json:1]: @symphony-elias what's your take on this? (putting all the mocked responses in the same file)

**Review by symphony-elias — APPROVED** [review]
LGTM !

**Pull request merged**